### ISSUE TITLE
How do you get an arbitrary model to walk around with animation?

### ISSUE THREAD

**Issue by arcman7** [flowtsohg/mdx-m3-viewer#70]
Demo [video](https://www.loom.com/share/f6efbd3de0ed42c185ae6c434b4fb5c8)

You can see in this video, the sheep models just float from one location to the next without any sort of walking animation. The code that updates their positions looks like this:
```
function walkSheepToNewLocations(directions, delta) {
      for (let i = 0; i < sEnv.num_sheep; i++) {
          sheepModelInstances[i].setSequence(sheepWalkSeq, true) // using that boolean extra arg introduced in PR https://github.com/flowtsohg/mdx-m3-viewer/pull/69
          sheepModelInstances[i].move([directions[i][0] * delta, directions[i][1] * delta, 0] )
      }
}
```

What's weird is I never came across this issue when animating a single Grunt model walking. I would merely set the sequence to the walk value and then start calling `gruntModelInstance.move([delta_x, delta_y, 0])`. 

With the sheep it doesn't matter if I remove `sheepModelInstances[i].setSequence(sheepWalkSeq, true)` from the for-loop or not, they still just float around.

**Comment by flowtsohg** [comment]
I don't think adding a boolean to the MDX handler is the correct solution - this is where the map viewer just doesn't give you what you want, so you'll have to make it do what you want.
For instance, if you want game-like functionality, you should probably have some basic implementation of units that have their own state, like the current order, and you base the animations on it.
This goes in the face of the existing automatic "just pick a stand animation" which is done [here](https://github.com/flowtsohg/mdx-m3-viewer/blob/master/src/viewer/handlers/w3x/map.ts#L572).
The more game-like functionality you want, you might realize it's not so easy to do, because this repo was designed to be a model viewer, not wc3, it's hard to do things that should be easy in the context of a game, but if you add some cool stuff I wouldn't mind a PR 😛
.

**Comment by arcman7** [comment]
Point me in the right direction and I'll do my best to act on it -
> For instance, if you want game-like functionality, you should probably have some basic implementation of units that have their own state, like the current order, and you base the animations on it.

What do you mean by the current order? 

My current implementation for adding a grunt unit looks like this:
```
async function addGruntUnit(pos = [0, 0, 0], rotation = [0,0,0,1]) {
  const model = window.gruntModel || await getGruntModel(false, false);
  const mockUnitInfo = { "location": [0,0,-1000],"rotation":rotation, "angle":rotation, "player":0,"scale":[1,1,1] };
  viewer.map.units.push(new Unit(viewer, model, {'comment(s)': 'Grunt'}, mockUnitInfo));
  let instance = model.addInstance();
  viewer.worldScene.addInstance(instance);
  instance.setLocation(pos);
  createPhysicsCylinder(25, 110, { kinematic: true, mdxM3Obj: instance }); // I know, don't judge
  return instance
}

class Unit {
  constructor(map, model, row, unit) {
    let instance = model.addInstance();
    instance.move(unit.location);
    instance.setRotation(unit.rotation);
    instance.scale(unit.scale);
    instance.setTeamColor(unit.player);
    instance.setScene(map.worldScene);
    if (row && row.moveHeight) {
      const heapZ = vec3.create();
      heapZ[2] = row.moveHeight;
      instance.move(heapZ);
      instance.setVertexColor([row.red / 255, row.green / 255, row.blue / 255, 1]);
      instance.uniformScale(row.modelScale);
    }
    this.instance = instance;
    this.row = row;
  }
}
```

**Comment by flowtsohg** [comment]
Consider a unit in a game - it's an object with a lot of state, like a position, facing, scale, health, mana, attack parameters, defense parameters, etc. etc. etc.,
Some of these things are somewhat incorporated into the viewer's own instances, but some isn't related to the viewer.
For example if you order a unit to move somewhere, you need to calculate the pathing needed, and ensure that this unit has its order, like "move" "smart" "attack" whatever, and it has the pathing needed, maybe also a target unit, and so on.
There are kinda endless details, but you can probably get something basic to work without too much issues, especially if you look at Warsmash, since it's basically a mish mesh of this repo, the HiveWE repo, and Reteras own modifications and additions.

I thought about adding some basic things like this to make the map viewer more interactive, but ultimately the fact that it runs so slow took away my motivation from working on the map viewer at all.
Years spent on optimizing this code and realizing that it will just NEVER run fast in JS kinda sucks.

**Comment by arcman7** [comment]
I'm looking at their repo to get a sense of what they did right now. 

> Years spent on optimizing this code and realizing that it will just NEVER run fast in JS kinda sucks.
What's your idea of slow? How many simultaneously moving units do you think could be rendered and still keep a decent frame rate?

**Comment by flowtsohg** [comment]
Slow as in you have a bunch of units and particles, not unlike any regular wc3 scenario while playing a game, and JS simply cannot hold 60FPS, and that's with only the graphics. Now on top of that add the Lua VM, add units with their orders and pathing and constant height tests, and so many more things, and it's just not going to hold.

I used to think Java is slow, but Retera's more-or-less copy of older worse code of mine is easily getting over 200FPS with scenarios the JS code can't get 30, and HiveWE is faster still since it's C++.

I spent countless hours optimizing this repo with many different types of optimizations, many of which no longer exist because they didn't help and made the code needlessly complex.
At the end of the day, I don't think it is possible to make real world fast code in JS, simply because of its memory model. No proper control over memory means that pretty much every access to every node in a skeleton, to...pretty much every single object in JS, will ultimately be a cache miss.
The optimizing engines in JS VMs can improve this a bit, but there's just no scalability here when you want something like a game engine.

**Comment by arcman7** [comment]
I believe a solution is currently being developed by Google (and every major web browser) for this specific class of problem:
https://youtu.be/K2JzIUIHIhc?t=976

Web GPU for chrome can be used currently on chrome canary
- [mac](https://www.google.com/chrome/canary/?platform=mac)
- [windows](https://www.google.com/chrome/canary/?platform=win64)

From there you need to enable web GPU:
chrome://flags/#enable-unsafe-webgpu
![Uploading Screen Shot 2021-04-19 at 6.26.02 PM.png…]()

[Github / more info](https://github.com/gpuweb/gpuweb/wiki/Implementation-Status)
[Specificationn](https://gpuweb.github.io/gpuweb/)

Am I wrong in thinking this solves the majority of the performance issues present in the mdx-m3-viewer codebase?

**Comment by arcman7** [comment]
Sorry I hope that wasn't presumptious, just got really excited when I saw this tech :D

**Comment by flowtsohg** [comment]
I thought in the past to convert the code to WASM, and I did get hyped about WebGPU when it was announced a long time ago, but will it matter much to this code base? I don't know.
Either way I doubt it's worth the effort at this point - if you want legit performance, the browser is not the right target platform.
I also don't have much time to work on code for the time being due to life stuff. :(

**Comment by arcman7** [comment]
Well that's fair. It would be a lot of work. I'm gonna try to see if I can get some of these models exported into a .babylon file format and try them out in their webGPU demo.

**Comment by arcman7** [comment]
> but will it matter much to this code base? I don't know.

I've got one data point for you-

Map - https://easyupload.io/kwcf1j
                      cells      instances     particles    fps
web GPU       13         653              ~1500        60
web GL          12         558              ~1400        32.5
![image](https://user-images.githubusercontent.com/9598327/116182724-9dc15180-a6d1-11eb-9cfc-05d3895e2a4c.png)

This is your own rendering code, unaltered, seeing an immediate benefit of double fps.

**Comment by flowtsohg** [comment]
I don't understand, if it's my code how does it run on WebGPU? 🤔

**Comment by arcman7** [comment]
> I don't understand, if it's my code how does it run on WebGPU? 🤔

I can't recall exactly how the API is set up, but I think it's meant to be backwards compatible. And in some cases should just start working -
https://www.construct.net/en/blogs/ashleys-blog-2/webgl-webgpu-construct-1519


To follow up on this, I'll disable the unsafe GPU flag and try it again. This will eliminate the possibility that the performance gains are simply the canary chrome browser being that much more optimized for general rendering operations.

**Comment by arcman7** [comment]
Can confirm, disabling the GPU with the flag `chrome://flags/#enable-unsafe-webgpu` brings the fps down to 32.5 on chrome canary

**Comment by flowtsohg** [comment]
So, I still don't understand if this is something I can run and how.

I don't doubt WebGPU can have better rendering performance than WebGL if the code was designed for it, and maybe it already runs somewhat faster also with a translation layer, although I don't quite understand what Construct is.

But, and this is an important but, this won't change the update stuff much.
Get a bunch of units visible, and performance is going to be bad due to node updates and such things.
WASM might improve that, but I somewhat doubt it will be that effective (and again, require A LOT of work).

**Comment by arcman7** [comment]
> But, and this is an important but, this won't change the update stuff much.
Get a bunch of units visible, and performance is going to be bad due to node updates and such things (even more so for Reforged models with their absurd amount of nodes).

Yes, I agree with your reasoning and to your point, I did a test wherein both canary and regular chrome I used that same map, zoomed as high as I could while still keeping everything in view (maxing the particle count), and watched the fps drop; 22 for canary 17 for regular.

When I have some time today I'm gonna read a bit more about webGPU to try and figure out if there's a viable path to an outcome that minimizes the object management done by JS.

**Comment by flowtsohg** [comment]
Particles could theoretically be sped up in two ways.

First, it needs to have a C-like memory model, an array of structs.
This is possible in JS by having a big shared buffer, where every consecutive N numbers represent a particle, and when moving a particle (e.g. a particle is born or died), rather than moving an object, the numbers need to move, much like a C++ move operation.

Second is to run the particles entirely in the shader.
This involves feedback transform with WebGL2 or whatever equivalent WebGPU offers.
The way it works is you store the shared buffer in a texture instead (e.g. texture buffer), and then you render this texture, with the output data being the updated particle data, which is then written to a framebuffer with a texture attached to it.
The next frame the input and output textures are swapped, and so on.

There are also simple instances, for which I did multiple experiments in the past, animation queries can be O(1) instead of O(n) which is a couple of % maybe, and many more things.

Yes, this code can become faster. But run the same exact code in C++ (HiveWE), or even in Java (Warsmash), and it just inherently works so much faster. At some point fighting the VM just to get some more juice in an ever losing battle becomes somewhat pointless.
That point pretty much became node updates.
Render 100 Mountain Kings and see how much effort is spent on updating the nodes.
And yes, 100 Mountain Kings is a bit odd, but 100 visible units is not even a lot for WC3, and it barely nudges it on the CPU side (it did used to have serious issues on the GPU side because it was old as heck, I think it's better since Reforged).

Compare the following images, one runs on the web and shows a small part of a map, the other from HiveWE rendering the same map in its entirety (it has TONS of objects and particles).
I don't quite know what HiveWE renders or not, but even if we disregard all emitted objects and whatnot...the frame times speak for themselves.
I also bothered showing my CPU usage, because the poor performance isn't the CPU
 dying from too much work. Rather stuff is very inefficient.

![Untitled](https://user-images.githubusercontent.com/5874825/116480463-0eab6980-a88a-11eb-83e6-068e77f28e5d.jpg)

![unknown](https://user-images.githubusercontent.com/5874825/116480481-1539e100-a88a-11eb-95c2-16a396381c13.png)

**Comment by arcman7** [comment]
> Second is to run the particles entirely in the shader.
This involves feedback transform with WebGL2 or whatever equivalent WebGPU offers.
The way it works is you store the shared buffer in a texture instead (e.g. texture buffer), and then you render this texture, with the output data being the updated particle data, which is then written to a framebuffer with a texture attached to it.
The next frame the input and output textures are swapped, and so on.
How you track particles that died, know where to add new particles in the buffer, and so on, I didn't think about enough to say off the top of my head.

 I don't really know enough yet at a low level how these models are rendered. 

Take for example [/src/parsers/mdlx/layer.ts](https://github.com/flowtsohg/mdx-m3-viewer/blob/a41a47f657be68d081f77d7a7fae96b452fb8979/src/parsers/mdlx/layer.ts)

It differs from some of the other layers.ts files in that it has `fresnel` parameters. How does this directly get used in the rendering of a model and then put into a shared buffer as a texture?

**Comment by arcman7** [comment]
Sorry, I've been reading through the files, and I still don't get the high level / low level concepts at play here. All I know at this point is that you write very clean code lol.

**Comment by flowtsohg** [comment]
Hmm, that's somewhat of a big question :P

There are meshes (geosets) with referenced materials.

Materials in TFT/SD work in layers sort of like how you'd have layers in 2D painting software, but the 3D equivalent.
For example, to get team colors, the first layer is the solid team color, so you get a fully red rendered mesh.
Then the next layer holds the actual mesh diffuse texture, which alpha blending, such that where the texture is transparent, you see the red below it.
This is extended to more combinations and blending operations that are supported.

Materials in HD always have a predefined structure with 6 layers that are all used for one draw command (which makes them simpler and faster, but support less exotic rendering).

Per-layer things like fresnel are used when drawing the layers - as GL settings, or inputs to the shaders (although fresnel specifically is an HD thing I didn't bother implementing).

Particles a quite different.
They hold only the per-particle information they need, like a location, and are updated globally by the scene (since particles need to update also if their instances don't get updated).
When rendering a specific emitter, all of its particles are iterated, the needed data is copied to one shared buffer, and this buffer is then used by the shader to render all of them.
If the design was different so that the particles ARE the buffer, we avoid constantly copying the data around, and data only needs to really "move" when a particle is created or removed.
Moving this entire process into the shader is a bit complicated, and can't be done with WebGL1 so I never looked too much into it.

**Comment by flowtsohg** [comment]
Just to illustrate what I mean by particles being the buffer - you can check out SkeletalNode, which works in the same way.
There is a shared buffer for the numerical data, and each node gets typed array references to this buffer.
So for example, all of the world matrices are held in one Float32Array, which is submitted in one go to the GPU when rendering, and each SkeletalNode gets its own Float32Array viewing a part of the shared one.
I cheated a little because the nodes then have non-shared stuff like parent references, object references, and the like.
In C++ these would also be simple numbers, I suppose, and could also be shared.

I set up a C++ WASM WebGL demo just to humor myself.
Maybe I'll attempt at making the smallest demo I can at rendering an animated MDX...we'll see. I do already have the required parsers in C++ that I wrote a while ago :P

**Comment by arcman7** [comment]
Okay, I'm sort of getting how this shared buffer is used, specifically with the skeletal nodes. What would be the easiest file type to dig into for a complete understanding from layers all the way up to it's object heirarchy?

**Comment by flowtsohg** [comment]
Particle systems typically emit simple quads that are billboarded so they face the camera no matter where you look at them from.
For wc3 particle emitters (which are ParticleEmitter2, more on this below), every time a particle is emitted, it needs to get some properties from the emitter based on the animation time of the owning instance, which would be the functions you listed above for ParticleEmitter.
In addition, every particle has its own life property, and some properties like the color are animated based on that. For instance, a fire particle might start blue-ish, and as it updates towards its death it might turn yellow-ish.

The handler has it split further.
Every model has e.g. a ParticleEmitter2Object which is the equivalent of the model object but with extra handling like loading textures and whatnot.
Then every instance has its own actual emitter (extending Emitter), which references the object in the model and does its things.

There are more emitters in wc3.
ParticleEmitter emits models, such as bones and guts flying all over the place when some units die.
ParticleEmitter2 emits quads that are typically billboarded, but can also be set to always face the XY plane (i.e. same orientation as the ground).
RibbonEmitter emits lines rather than quads, and all of the existing ribbons are chained together to form a mesh. 1st ribbon and 2nd ribbon are connected to form a quad, 2nd and 3rd are connected, and so on. The texture it uses is stretched across the entire mesh.
EventObjects are a group of emitters that work slightly differently, but ultimately emit either models (e.g. illidan's burning footsteps), splats which are quads flat on the terrain (i.e. follow terrain deformations), ubersplats which are like splats (any difference? 🤔 ), and sounds which are used for footsteps, death sounds and such.


On the WASM note, I started experimenting with it a bit. So far the environment is horrible. I can't get Visual Studio to understand any WASM project I tried without manually adding every path and option needed, because the given CMake files don't work correctly.
Either way, I did manage to get stuff going by manually compiling in the command line. It's very slow and annoying, but it works for now.
The real problem that I have for now is that allocations seem to be incredibly slow on the WASM side for some reason.
i.e. trying to load Footman.mdx (~105KB) would take 10 seconds when I was copying the buffer from JS to a newly allocated buffer in WASM.
Passing the buffer in a more hacky way and avoiding the WASM-side buffer allocation made it load significantly faster, but it's still relatively very slow, most likely because of allocations in the MDLX parser itself.
To be more specific, using C functions like malloc/free is extremely fast. Creating a new std::vector with a size (or reserving/resizing) is extremely slow.

**Comment by arcman7** [comment]
The particle system makes sense, I believe. I'll have to test that understanding somehow. 

> Creating a new std::vector with a size (or reserving/resizing) is extremely slow.

Is std::vector object is a contiguous object or not? Is it possible to use arrays of bytes as the buffer?

I really don't get why allocating space to an std::vector would be slow.

**Comment by flowtsohg** [comment]
I started writing information about how to handle MDX models in the [specs](https://www.hiveworkshop.com/threads/mdx-specifications.240487/), but I ended up never quite adding all of the information I wanted. There is so much of it, it could fill a whole book chapter. Then again, I am not sure how many people would benefit from that, since programmers will eventually look at the code anyway, and it's not so complex nowadays (not so much in v4, now that was hard to read!), plus there are multiple implementations to look at.

I have no idea why std::vector is so slow either.
I did a small change and moved to Rust instead :P
It will take some time to learn the language properly, but at the very least getting a demo with WebGL running took a few minutes.
![Untitled](https://user-images.githubusercontent.com/5874825/116949642-895df580-ac8b-11eb-825e-7ffc8ad9765b.jpg)

**Comment by arcman7** [comment]
Woah. How does that work?? You can run RUST from the browser the same way you can run WASM?

**Comment by flowtsohg** [comment]
It also has the tools to compile to WASM and have interoperability.
I think it can be compiled directly with Emscripten, and has GLFW bindings and the like as well.
But...it has its own native tools that support this with no extra work.
All I did was take the [WebGL example](https://github.com/rustwasm/wasm-bindgen/tree/master/examples/webgl) and modify it a bit.
I suppose the next step is to get an ArrayBuffer in, and see how to write a parser.

**Comment by arcman7** [comment]
That is amazing. RUST really sounds like the best solution. Here's me crossing my fingers for good performance 🤞

**Comment by arcman7** [comment]
Is there a link to the workflow setup you're using with RUST? Just curious to see it.

**Comment by flowtsohg** [comment]
I tried other demo projects before, so I am not 100%, but I think all you need to do is install Rust which comes with Cargo (Rust's NPM), download the demo from the link above, and run `cargo install`.
For the NPM part obviously have NPM installed, and run `npm run serve` to run in dev-watch mode.

Meanwhile I am stuck at understanding how one is supposed to parse binary data :P

**Comment by flowtsohg** [comment]
It's not much, but I finally get how to read stuff. No time to do with it anything for now though.

```rust
pub fn load_file(&self, buffer: Box<[u8]>) -> u32 {
    let mut cursor = Cursor::new(&buffer);
    
    let mut mdlx = String::from("AAAA");
    unsafe {
        cursor.read(mdlx.as_bytes_mut()).expect("NOPE");
    }

    let mut vers = String::from("AAAA");
    unsafe {
        cursor.read(vers.as_bytes_mut()).expect("NOPE");
    }

    let size = cursor.read_u32::<LittleEndian>().unwrap();
    let version = cursor.read_u32::<LittleEndian>().unwrap();

    version
}
```

**Comment by arcman7** [comment]
That's super nifty, I'm gonna have to try that out soon! Probably Thursday. 

> console.log(v.load_file(new Uint8Array(buffer))); // 800

So was there any performance gain using rust to allocate buffers?

**Comment by flowtsohg** [comment]
Allocations are still slow, but not 10 seconds slow.

I am not sure if it's possible  to share any kind of arrays from Rust to JS though.
This includes strings, primitive arrays, vectors, whatever.
Strings can be cloned in and out. Not ideal, but it is how JS works anyway.
I am not sure how I could share vectors, matrices, vertices, etc. though.
I probably shouldn't bother with making things accessible in JS and just write only what I need to render stuff, but it's somewhat discouraging that WASM is so...restricted (this has nothing to do directly with Rust, but rather with how WASM has no knowledge of memory and only JS controls it).

**Comment by flowtsohg** [comment]
I'll be honest, after more playing and figuring what isn't really possible...I am becoming quite skeptic about WASM.
Its memory limitation is becoming more obvious as I try to write more things.
Not being able to use any strings or vectors, and therefore any structs that have any string or vector fields, and so on...
This of course doesn't say anything about runtime performance, but how is one even supposed to make a real-world library that handles data with these limitations, I am not sure.

**Comment by arcman7** [comment]
> Not being able to pass back to JS any shared strings or vectors, and therefore any structs that have any string or vector fields, and so on...

Do you mean that literally or is it just really slow?

Also, are you pushing this scratch work up anywhere? I wouldn't mind being able to see it - maybe profile it a bit. I have a roommate who happens to be pretty good at optimizing stuff like that.

**Comment by flowtsohg** [comment]
Let me rephrase that. You can pass whatever you want, since "passing" means mostly moving bytes from context to context.
But wasm_bindgen (Rust's native tool) doesn't support structs with dynamic memory, only ones that it can essentially do a `memcpy` to pass.
Since strings, vectors, etc. are heap allocated, with the way Rust works they can't be `memcpy`'d and must be cloned for a legit deep copy.
For strings at the top, like `MdlxModel.name`, this can be solved by not exposing them directly, but rather writing a getter/setter:
```rust
#[wasm_bindgen]
impl MdlxModel {
    #[wasm_bindgen(getter)]
    pub fn name(&self) -> String {
        self.name.clone()
    }

    #[wasm_bindgen(setter)]
    pub fn set_name(&mut self, s: String) {
        self.name = s;
    }
}
```
But let's look at `MdlxSequence`:
```rust
#[wasm_bindgen]
pub struct MdlxSequence {
    name: String,
    pub start: u32,
    pub end: u32,
    pub move_speed: f32,
    pub flags: u32,
    pub rarity: f32,
    pub sync_point: u32,
    pub extent: MdlxExtent,
}
```
It also has a string name, so we make a getter/setter and it works. Right? No.
This would allow to instantiate an `MdlxSequence` on the JS side, the bindings will be correct.
However we still can't actually RETURN any `MdlxSequence` from the WASM side, because it can't be copied.

And then there's the real crux - how can WASM support vectors? even if there was a stack allocated vector and it could be `memcopy`'d, there is no mechanism to synchronize changes on the WASM and JS sides.
What if one side pushes a new object, or changes the order. These changes won't be reflected on the other side.
Maybe this can be solved with glue code (similar to Vue?), but it's becoming more and more messy.

There isn't enough code to really show much yet, I don't have a lot of time, and most of it was spent on understanding Rust and trying to figure how to use wasm_bindgen.

**Comment by flowtsohg** [comment]
Pushed a bunch of changes I had stacking up over a long time, one of them is a simple example of how to add state to the units and doodads so you don't have to do weird hacky things to not get the automated stand animations.

For example if you have your sheep as a `Unit`, you can do `sheep.state = WidgetState.WALK` (from [widget.ts](https://github.com/flowtsohg/mdx-m3-viewer/blob/master/src/viewer/handlers/w3x/widget.ts#L6)) and it won't run the stand animations.

**Comment by arcman7** [comment]
Awesome! Looking now. I just had the second dose of the Pfizer vaccine and I was completely out of commission these past few days. Excited to bounce back and try this out!

**Comment by arcman7** [comment]
Definitely gonna try a reforged map on chrome canary's webGPU :D

**Comment by flowtsohg** [comment]
I am not sure if it will actually load a Reforged map, I guess my comment wasn't very explanatory. It just supports loading all of the base game files using the new structure :P

Although technically I think the files used by the map viewer should support Reforged? I don't think I ever tried.

The API doesn't have TFT files unfortunately. You'll have to use the old url that gets files from an unpacked game.
Note that the API is much more than a simple GET, aside from supporting SC2 (and I think some other Blizzard stuff?) it does the whole hierarchy thing for WC3, and I believe fully supports overriding the root as you can do in the Reforged WE, such as starting a path with `_hd.mod:` (or something similar, I don't quite remember). Unfortunately it still doesn't support a tileset parameter, which is why the map viewer doesn't render the right cliff textures.

**Comment by arcman7** [comment]
> Second there's the question of support. When I started this viewer, it was always using the newest slickest features on the web as they came online. Support was poor for different features, and I had to implement some things manually for different browsers, and user browsers supported or didn't support all sorts of things. It was a mess. Exciting, but a mess. Do I want to go that route now after 10 years of constantly fixing and upgrading the same code over and over and over and ensuring it works on as many devices while still being mostly updated? I don't know. That's why I never updated to WebGL2 after all.

I am almost certain that webGPU has the full support of Safari, Mozilla, new Edge (chromium), and ofc Chrome. There shouldn't be much mess as far as that API goes. And yes, webGPU won't magically make everything better, but considering that it does have better performance with your code, it would be the place to test reforged asset rendering.

**Comment by flowtsohg** [comment]
I don't mean whether it has the support of the browser vendors, rather is it actually supported right now for the average user using their browser, and the answer to that is no.
It will be in the future probably, as vendors complete their implementations and remove the flags that hide them, and as people circulate whatever old OpenGL1/2 capable phones they still have.

If the code could be decently rewritten to somehow support both APIs it would be great, but I don't know enough about WebGPU or have much time or motivation for the time being to do much.

**Comment by arcman7** [comment]
Oh, okay, that makes more sense now. I see why you're more interested in exploring the WASM/RUST alternative for the time being. However, once it is released publically, it could make a huge impact on the performance here... if I were to get as far as porting over a few basic shaders, connect it into the current setup with some sort of webGPUIsSupporeted flag, and show you what this would look like in a feature branch, would that be something you'd be interested in seeing?

**Comment by flowtsohg** [comment]
I think that the shaders are the least of the concerns when it comes to WebGPU. WebGL shaders can be compiled to it, and it's not hard to make properly compiled binaries as well.
My concern is...all the rest.
Everywhere that WebGL is used will be to be changed in some way.
Does this mean an entirely separate handler? perhaps branching on every piece of code that calls GL? perhaps isolate all of the GL state more (most of it is somewhat isolated already) and then add branching on specific methods but keep the handlers more or less the same. Maybe something else. I really don't know enough about WebGPU/Vulkan to say how different the code would be. It seems like it will mostly let move most of the runtime calls to be instead at setup, which would be nice. I don't know how dynamic stuff like bone textures or particle buffers would work.
I don't really have time to work on the Rust thing either :/

**Comment by arcman7** [comment]
>  I really don't know enough about WebGPU/Vulkan to say how different the code would be.

Currently, in the process of relocating to Texas, I'll have a branch up shortly after where I'll start trying to see what that looks like.

**Comment by arcman7** [comment]
Hey, back on this again lol. Who knew relocating and starting a new job could take up so much time 😄

**Comment by flowtsohg** [comment]
Welcome back! Not much changed around here though :P

**Comment by arcman7** [comment]
>  Bigger update
> - The clients now get built with webpack. I might split it to two different webpack configs, still not sure about the setup.
> - Added the downgrader client, to make Reforged maps openable in old editors. It doesn't copy triggers, and doesn't support all of the Reforged files (see the issues).
> - Copied Retera's Warsmash handling of axis-specific billboarding for skeletal nodes.
> - Added a tiny bit Reforged stuff to the Jass2 context. But really just a tiny bit.

 I see some new commits! 
 
 Also, I realized that that the choppy animation I was experiencing was not an issue of mdx-m3-viewer's runtime speed... It looks like something else is causing the issues. I tried a different model (a brown wolf), and also tried bringing the number of model instances down to < 5, and the choppiness was still there. I need to figure out why/how I'm doing this wrong.

**Comment by flowtsohg** [comment]
I updated lots of small things, but nothing major changed or probably will change(?)

Are you running on a non-60Hz monitor? if so, make sure to control `dt` when calling `viewer.update()` or `viewer.updateAndRender()`, see the last section of the README.

**Comment by arcman7** [comment]
> Variable frames per second
> ModelViewer.update() and ModelViewer.updateAndRender() have an optional dt argument.
> 
> dt controls how much time in miliseconds to advance the animations.
> 
> By default, dt is set for 60FPS, or 1000 / 60.

How can you actually match the rate at which `ModelViewer.update()` get's called?

`let fps = 1000/60;`
If I want to run the animations at `fps` that would mean waiting for the duration of `fps` before calling update again. So if I want to advance the animations by `fps`, how can I be assured that from the javascript event-loop these update functions also get called at a frequency of `fps`?

**Comment by flowtsohg** [comment]
To clarify, the idea is not to set the FPS of the viewer, but rather match the FPS of the browser.

The recommended thing in JS is to use requestAnimationFrame, which runs at the monitor's refresh rate.

Since JS controls the actual waiting, we can check what the time is compared to last frame, and see what the FPS is like that, and if you know the FPS you can control the animation speed.

You could of course pass any `dt` you want if you want to affect time, and there's also `ModelInstance.timeScale` to affect the animation of specific instanes (although this isn't relevant to the issue).

**Comment by arcman7** [comment]
Okay, I understand now what you're saying regarding the fps. Thank you for the clarification there.

As for the animations, I noticed the model instances are having their animation sequence being reset by something; probably similar to what was happening here 
https://github.com/flowtsohg/mdx-m3-viewer/pull/69/files#

Did you revert any of that stuff? I don't see how I can set an animation sequence on a model instance and have it be preserved throughout the map update calls right now.

**Comment by flowtsohg** [comment]
Look a couple of message up where I added `WidgetState.WALK` :P

**Comment by arcman7** [comment]
> For example if you have your sheep in map.units, you can do sheep.state = WidgetState.WALK (from widget.ts) and it won't run the stand animations.

😅 
Omg. I feel dumb. It's been too long! I'm just surprised I didn't immediately change my code to use those changes.
